### ISSUE TITLE
`prefer-const` reports variables initialized in another function scope

### ISSUE THREAD

**Issue by erikdesjardins** [eslint/eslint#5251]
ESLint 2.0.0, npm 3.3.6, node 5.0.0.

Config:

```json
{
	"parserOptions": {
		"ecmaVersion": 6
	},
	"rules": {
		"prefer-const": 2
	}
}
```

Code:

```js
let foo;
function bar(baz) {
	foo = baz;
}
```

ESLint reports:

`1:5  error  'foo' is never modified, use 'const' instead  prefer-const`

But `foo` couldn't be `const` because `baz` isn't known immediately.

IIRC this worked correctly (i.e. didn't report) in ESLint 1.10.3.

**Comment by ilyavolodin** [comment]
Hmm.. That does look like a bug, since you could do:
```js
bar(22);
bar('abc');
```
@mysticatea could you confirm for me that this is an oversight?

**Comment by mysticatea** [comment]
In this case, the `foo` variable is not never used outside of the `bar` function. So `prefer-const` considers the variable can be moved to the function.
The following patterns are not warned.

```js
let a;
function foo(b) {
    a = b;
}
bar(a);
```

```js
/*exported a*/
let a;
function foo(b) {
    a = b;
}
```

```js
function foo(b) {
    const a = b;
}
```

**Comment by erikdesjardins** [comment]
Huh, you're right.

I was seeing this in quite a few places in our code, but when I look closer it seems that, for every single occurrence, the var *could* be moved inside the function.

Sorry about that!

**Comment by mysticatea** [comment]
Hm, this behavior may be good to split out to something like new `prefer-small-scope` rule. I noticed this rule is doing two things, "detecting it's never modified" and "detecting it can be moved".

**Comment by mdebruijne** [comment]
@mysticatea, question; this bug and duplicates are closed by the reporter, but is this the expected behavior of prefer-const? In the examples you and others provided you clearly cannot replace let with const without breaking the code.

If this is not the expected behavior of prefer-const then would you be so kind to reopen this issue report?
If this is the expected behavior of prefer-const then would you be so kind to add some notes to the documentation (http://eslint.org/docs/rules/prefer-const)? There is no information in the documentation that the behavior differs depending on the scope of the variable.

**Comment by mysticatea** [comment]
@mdebruijne This was the intentional behavior of `prefer-const`. The `foo` can be moved into the `bar` function safely in this case. But I have a plan to delete the behavior and split out to `prefer-smaller-scope` rule (#5284).

**Comment by mdebruijne** [comment]
@mysticatea; thanks for clarifying, nice rules, much appreciated.
